### PR TITLE
tblastx filtering accounting for reading frames

### PR DESCRIPTION
**Pull request by msarrias**
## The criteria for filtering the` tblastx` hits are the following:

1. **Minimum Query Alignment Coverage:** Retain hits that have at least 90% query alignment coverage.
2. **Reading Frame:** Only keep hits where the reading frame of the query CDS matches the hit query reading frame.
3. **Handling Overlaps with Preference:** For hits sharing the same query CDS and having overlapping target coordinates:
- Favor hits that are in the same strand as the gene. 
- Among those, prioritize hits with the lowest e-value.

## Extra info for understanding CDS genomic coordinates and reading frames in GFF files:

Consider a mRNA sequence composed of several CDSs.

**genomic coordinates**
```
mRNA_1
cds1_1: (0,127)      reading frame: 0
cds2_1: (4545,4682)  reading frame: 2 --> the initial two nucleotides of cds2 complete the last codon of cds1
cds3_1: (6460,6589)  reading frame: 0
cds4_1: (7311,7442)  reading frame: 0
```

If we now account for the reading frames, the CDS coordinates adjust to:
**transcriptomic coordinates**
```
mRNA_1
cds1_1: (0, 127), (4545, 4547) 
cds2_1: (4547, 4682)           
cds3_1: (6460, 6589)           
cds4_1: (7311, 7442)   
```        
Now, let's consider a second mRNA transcript associated with the same gene:
**genomic coordinates**
```
mRNA_2
cds1_2: (0,127)      reading frame: 0
cds3_2: (6460,6589)  reading frame: 2
cds4_2: (7311,7442)  reading frame: 2
```
**transcriptomic coordinates**
```
mRNA_2
cds1_2: (0, 127), (6460, 6462)         
cds3_2: (6462, 6589),(7311, 7313)      
cds4_2: (7313, 7442)   
```  

**Take away message:**

- When considering different CDSs that share genomic coordinates, if the CDSs are affected by alternative splicing events, they might exhibit differences in their transcriptomic sequences. (For obligate cases, it would be interesting to check the reading frames of the CDSs in question, as they might yield different sequences at the protein level).

**On another note...**

Another thing that has me confused is that if the transcript sequence is not divisible by three, it is always the case (at least for humans) that the last CDS has some extra nucleotides.  I really don't understand this. (wild guess) Is this because those extra nucleotides would account for the "completion" of other mRNA transcripts?  By saying this I'm assuming that the last CDS is always present in all transcripts and therefore has the task of "completing" preceding CDSs. A half-baked attempt to convince myself that those loose nucleotides accomplish a function.

The previous is a good example of this:
```
mRNA1
cds1_1: (0, 127), (4545, 4547) length: 127 + 2  % 3: 0
cds2_1: (4547, 4682)           length: 135 % 3: 0
cds3_1: (6460, 6589)           length: 129 % 3: 0
cds4_1: (7311, 7442)           length: 131 % 3 : 2 
```

```
mRNA2
cds1_2: (0, 127), (4545, 4547) length: 129 % 3: 0
cds3_2: (6462, 6589),(7311, 7313)                length: 129 % 3: 0
cds4_2: (7313, 7442)           length: 129 % 3 : 2 
```

### CODE REVIEW

File: exonize/sqlite_utils.py [hunk 1]
```diff
@@ -645,68 +647,81 @@ def create_filtered_full_length_events_view(db_path: str, timeout_db: int) -> No
...
-        WITH filtered_overlapping_queries AS (
-        WITH overlaping_fragments AS (
-        WITH gene_id_counts AS (
-        SELECT f.gene_id, f.CDS_start, f.CDS_end
-        FROM Fragments AS f
-        WHERE f.percent_query >= 0.9
-        GROUP BY f.gene_id, f.CDS_start, f.CDS_end
-        HAVING COUNT(*) > 1)
+        WITH
+        -- Identify candidate fragments that satisfy our coverage and in-frame criteria.
```
Comment (msarrias) [inline exonize/sqlite_utils.py:651]: @arvestad Could you please check this view. Thanks!

Comment (arvestad) [reply exonize/sqlite_utils.py:651]: This change is too big for me to be able to review within GitHub. I will accept the PR and look in the source.

**Comment by arvestad** [comment]
Won't all all genes with a duplicated fragment become a "multi fragment gene"? If there is an exon _e1_ that has been duplicated to _e2_, then both _e1_ and _e2_ will be used as queries and hit a "target fragment" (_e1_ hits _e2_, and _e2_ hits _e1_) to end up in the table _in_frame_candidate_fragments_. Then, the following query will pick up the gene, even though it is "only" a single duplication. Is this a bug or a feature?

```
        -- Identify gene_ids with more than one dupl fragment
            multi_fragment_genes AS (
	        SELECT cf.*
        	FROM in_frame_candidate_fragments AS cf
        	GROUP BY cf.gene_id, cf.CDS_start, cf.CDS_end
        	HAVING COUNT(*) > 1
		),
```

**Comment by msarrias** [comment]
We take care of that in the reconciliation step where we assign a "pair ID" to those cases i.e., the hit and the reciprocal hit. Do you think it could be handled differently?

**Comment by arvestad** [comment]
I am just trying to understand and wanted to make sure this was your intention.

10 okt. 2023 kl. 10:51 skrev Marina Sarrias ***@***.***>:



We take care of that in the reconciliation step where we assign a "pair ID" to those cases i.e., the hit and the reciprocal hit. Do you think it could be handled differently?

**Comment by msarrias** [comment]
Of course, the reason I did it this way was to handle cases like the following:

```
fragment_id  gene_id          gene_start  gene_end     CDS_start  CDS_end    query_start  query_end  target_start  target_end  
-----------  ---------------  ----------  ---------    ---------  ---------  -----------  ---------  ------------  ----------  
84820        ENSG00000188517  108808724   109302752       29373     29454     0            81         37414         37495       
84909        ENSG00000188517  108808724   109302752       37414     37495     0            81         29373         29454
84910        ENSG00000188517  108808724   109302752       37414     37495     3            81         29376         29454
```

You can see that both 84909 and 84910 can be the reciprocal match of 84820 (or the other way around).

The pair ID is assigned by the "assign_pair_ids" method defined within the Exonize class.

**Pull request merged**